kolibri/core/auth/management/commands/bulkimportusers.py

```diff
@@ -0,0 +1,587 @@
+
+    def add_check(self, header_name, check, message):
+        """
+        Add a header check, i.e., check whether the header record is consistent
+        with the expected field names.
+        `header_name` - name of the header for the column to be checked
+        `check`- function to be used as validator of the values in the column
+        `message` - problem message to report if a value is not valid
+        """
+        self._checks.append((header_name, check, message))
+
+    def get_username(self, row):
+        username = row.get("Username")
+        if username in self.users.keys():
+            return None
+
+        return username
+
+    def check_classroom(self, row, username):
+        def append_users(group, key):
+            try:
+                groups = row.get(key, None).split(",")
+                for classroom in groups:
+                    if not classroom:
+                        continue
+                    if classroom in group:
+                        group[classroom].append(username)
+                    else:
+                        group[classroom] = [
+                            username,
+                        ]
+            except AttributeError:
+                # there are not members of 'key'
+                pass
+
+        # enrolled learners:
+        append_users(self.classrooms, "Enrolled in")
+
+        # assigned coaches
+        user_role = row.get("User type", "learner").lower()
+        if user_role != "learner":
+            # a student can't be assigned to coach a classroom
+            append_users(self.coach_classrooms, "Assigned to")
+            self.roles[roles_map[user_role]].append(username)
+
+    def validate(self, data):
+        """
+        Validate `data` and return an iterator over errors found.
+        """
+        for index, row in enumerate(data):
+            error_flag = False
+            username = self.get_username(row)
+            if not username:
+                error = {
+                    "row": index + 1,
+                    "message": "Duplicated username",
+                    "field": "Username",
+                    "value": row.get("Username"),
+                }
+                error_flag = True
+                yield error
+
+            for header_name, check, message in self._checks:
+                value = row[header_name]
+                try:
+                    check(value)
+                except ValueError:
+                    error = {
+                        "row": index + 1,
+                        "message": message,
+                        "field": header_name,
+                        "value": value,
+                    }
+                    error_flag = True
+                    yield error
+                except Exception as e:
+                    error = {
+                        "row": index + 1,
+                        "message": "Unexpected error [%s]: %s"
+                        % (e.__class__.__name__, e),
```

Comment:
very minor note: generally I'm a fan of `string.format()` rather than printf-style string formatting because it looks more pythonic